src/daft-csv/src/read.rs

```diff
@@ -199,6 +220,387 @@ fn tables_concat(mut tables: Vec<Table>) -> DaftResult<Table> {
     )
 }
 
+#[derive(Debug)]
+struct CsvBufferPool {
+    buffers: Mutex<Vec<Vec<read::ByteRecord>>>,
+    buffer_size: usize,
+    record_buffer_size: usize,
+    num_fields: usize,
+}
+
+struct CsvBufferPoolRef<'a> {
+    pool: &'a CsvBufferPool,
+    buffer: Vec<read::ByteRecord>,
+}
+
+impl CsvBufferPool {
+    pub fn new(
+        record_buffer_size: usize,
+        num_fields: usize,
+        chunk_size_rows: usize,
+        initial_pool_size: usize,
+    ) -> Self {
+        let chunk_buffers = vec![
+            vec![
+                read::ByteRecord::with_capacity(record_buffer_size, num_fields);
+                chunk_size_rows
+            ];
+            initial_pool_size
+        ];
+        CsvBufferPool {
+            buffers: Mutex::new(chunk_buffers),
+            buffer_size: chunk_size_rows,
+            record_buffer_size,
+            num_fields,
+        }
+    }
+
+    pub fn get_buffer(&self) -> CsvBufferPoolRef {
+        let mut buffers = self.buffers.lock().unwrap();
+        let buffer = buffers.pop();
+        let buffer = match buffer {
+            Some(buffer) => buffer,
+            None => {
+                vec![
+                    read::ByteRecord::with_capacity(self.record_buffer_size, self.num_fields);
+                    self.buffer_size
+                ]
+            }
+        };
+
+        CsvBufferPoolRef { pool: self, buffer }
+    }
+
+    fn return_buffer(&self, buffer: Vec<read::ByteRecord>) {
+        let mut buffers = self.buffers.lock().unwrap();
+        buffers.push(buffer);
+    }
+}
+
+// Daft does not currently support non-\n record terminators (e.g. carriage return \r, which only
+// matters for pre-Mac OS X).
+const NEWLINE: u8 = b'\n';
+const DEFAULT_CHUNK_SIZE: usize = 4 * 1024 * 1024; // 1MiB. TODO(desmond): This should be tuned.
+
+/// Helper function that finds the first new line character (\n) in the given byte slice.
+fn next_line_position(input: &[u8]) -> Option<usize> {
+    // Assuming we are searching for the ASCII `\n` character, we don't need to do any special
+    // handling for UTF-8, since a `\n` value always corresponds to an ASCII `\n`.
+    // For more details, see: https://en.wikipedia.org/wiki/UTF-8#Encoding
+    memchr::memchr(NEWLINE, input)
+}
```

Comment:
won't this break on quoted columns with a newline in them? 

```csv
c0,c1,quoted_column
1,true,"this is \n quoted"
```

Reply:
Yeah that's correct. I left some followup TODOs below, but we haven't handled this case yet.

Some ideas we tossed around:
- Do speculative parsing (assume each chunk could start in quoted or unquoted case, then prune the possibility as we go)
- Check if we're in a file without quoted fields/let users specify this. In the happy path without quoted fields we can just rip through it. If not use speculative parsing.
- We can even imagine other edge cases that we might not handle correctly. In those cases simply fallback to an existing mature csv reader.